### PR TITLE
expr: Refine decimal type infer

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #30961

Problem Summary:

### What is changed and how it works?
Refine type infer for decimal multiply.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -121,7 +121,11 @@ func setFlenDecimal4RealOrDecimal(ctx sessionctx.Context, retTp *types.FieldType
 		if isMultiply {
 			digitsInt = a.GetFlen() - a.GetDecimal() + b.GetFlen() - b.GetDecimal()
 		}
-		retTp.SetFlenUnderLimit(digitsInt + retTp.GetDecimal() + 1)
+		if isMultiply {
```
Comment (hawkingrei) [inline expression/builtin_arithmetic.go:124]: It has two ```isMultiply``` checks. Can we merge it ?

Comment (windtalker) [reply expression/builtin_arithmetic.go:120]: Good idea, done.

**Review by wshwsh12 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**